beluga/include/beluga/utility/indexing_iterator.hpp

```diff
@@ -0,0 +1,205 @@
+  [[nodiscard]] difference_type operator-(const IndexingIterator& other) const noexcept {
+    if (cursor_ < other.cursor_) {
+      return -static_cast<difference_type>(other.cursor_ - cursor_);
+    }
+    return static_cast<difference_type>(cursor_ - other.cursor_);
+  }
+
+  /// Dereferences iterator at a given `offset` from its current position.
+  /**
+   * Behavior is undefined for offsets that take the iterator past the limits of the target container.
+   */
+  [[nodiscard]] reference operator[](difference_type offset) const noexcept { return (*indexable_)[cursor_ + offset]; }
+
+  /// Dereferences iterator at its current position.
+  /**
+   * Behavior is undefined if the iterator is past the limits of the target container.
+   */
+  [[nodiscard]] reference operator*() const noexcept { return (*indexable_)[cursor_]; }
+
+  /// Dereferences iterator at its current position and yields a pointer to it.
+  /**
+   * Behavior is undefined if the iterator is past the limits of the target container.
+   */
+  [[nodiscard]] pointer operator->() const noexcept { return std::addressof((*indexable_)[cursor_]); }
+
+  /// Checks if iterator position is strictly before that of another.
+  /**
+   * Can only be true for iterators that have the same target container.
+   */
+  bool operator<(const IndexingIterator& other) const noexcept {
+    return indexable_ == other.indexable_ && cursor_ < other.cursor_;
+  }
+
+  /// Checks if iterator position is before or equal to that of another.
+  /**
+   * Can only be true for iterators that have the same target container.
+   */
+  bool operator<=(const IndexingIterator& other) const noexcept {
+    return indexable_ == other.indexable_ && cursor_ <= other.cursor_;
+  }
+
+  /// Checks if iterator position is strictly after that of another.
+  /**
+   * Can only be true for iterators that have the same target container.
+   */
+  bool operator>(const IndexingIterator& other) const noexcept { return !((*this) <= other); }
+
+  /// Checks if iterator position is after or equal to that of another.
+  /**
+   * Can only be true for iterators that have the same target container.
+   */
+  bool operator>=(const IndexingIterator& other) const noexcept { return !((*this) < other); }
+
+  /// Checks if iterator position is equal to that of another.
+  /**
+   * Can only be true for iterators that have the same target container.
+   */
+  bool operator==(const IndexingIterator& other) const noexcept {
+    return indexable_ == other.indexable_ && cursor_ == other.cursor_;
+  }
+
+  /// Checks if iterator position is not equal to that of another.
+  bool operator!=(const IndexingIterator& other) const noexcept { return !(*this == other); }
+
+ private:
+  Indexable* indexable_{nullptr};
+  Index cursor_{};
+};
+
+/// Rewinds `iterator` a given `offset`, yielding a modified copy.
+template <class Indexable, class Index>
+[[nodiscard]] IndexingIterator<Indexable, Index> operator+(
```

Comment:
Implementing this as a hidden friend so it is only found by ADL would be much nicer to the compiler (and error messages for unrelated operator+ overloading errors).

Reply:
Wish we could do the same for the enum pipe operator overloads.

Reply:
Good idea. Done in https://github.com/Ekumen-OS/beluga/pull/320/commits/d451b5add74832b2397a3625c1a58d4565b409eb.

Reply:
> Wish we could do the same for the enum pipe operator overloads.

Do we need to? I mean, other than because it looks cleaner. Enum classes cannot derive from each other and are not convertible to each other unless explicit provisions are put in place.

Reply:
See how the enum operator pipe overload is listed as a candidate when `ranges::to` fails to compile. That's unfortunate. To be fair, a couple of overloads in the std namespace are listed as well, but not that many.

Reply:
I see. Yeah, I agree it adds unnecessary clutter. I guess we could also drop the strongly typed enums and use plain integer flags. I was on the fence about that too.

Reply:
Range-v3 does something very similar with range categories: https://github.com/ericniebler/range-v3/blob/97452bb3eb74a73fc86504421a6a27c92bce6b99/include/range/v3/view/any_view.hpp#L43C16-L43C24

I think it's fine the way it is.